### ISSUE TITLE
 Null / NaN values in bin-ordinal scale

### ISSUE THREAD

**Issue by kanitw** [vega/vega#997]
Given a Vega-Lite spec,

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "description": "A scatterplot showing horsepower and miles per gallons for various cars.",
  "data": {"values": [
        null,
        "a",
        22,
        30,
        48
      ]},
  "mark": "point",
  "encoding": {
    "y": {"field": "data","type": "ordinal"},
    "color": {"bin": true, "field": "data","type": "ordinal"}
  }
}
```

which will produces [this Vega spec](https://pastebin.com/raw/yLFSHpEx) and the following plot

![image](https://user-images.githubusercontent.com/111269/29578891-cbcc7bc4-8725-11e7-92be-576e7db953ad.png)

Bin transform seems to keep `null` / `NaN` as is. 
However, the bin-ordinal scale does not handle null / NaN values so the color does not appear and the legend incorrectly includes 0.  

I can easily change `stroke` to use a production rule:
```
          [{
            "test": "datum.data === null || isNaN(datum.data)",
            "value": "grey"
          },{
            "scale": "color",
            "field": "bin_maxbins_6_data_start"
          }]
```

![image](https://user-images.githubusercontent.com/111269/29578876-bdeab35e-8725-11e7-9c08-6a5b4b800a5f.png)

to make the color appear, but that won't make the legend correct.  Thus, I wonder if there is a workaround for this.  

One hack that I was thinking is to 
(1) set `legend.values` to a concatenation of scale's domain and `[null,NaN]`
(2) apply production rules to the legend's encode block to similarly use grey for `null` /`NaN`

However, seems like there is no array concat in Vega expression.  Plus, I wonder if some of these null handling logic should be in Vega.  

cc: @akshatsh -- This is a relevant experiment based on what I try when reviewing your code.

**Comment by kanitw** [comment]
From a further experiment, a part of the problem is that the extent transform put `null` as the min value of the extent.  

If I replace bin `extent` with `[22,48]`, then at least the legend  works correctly with non-null values:

![image](https://user-images.githubusercontent.com/111269/29643315-2831508a-8823-11e7-9d68-3fad4e09dcc0.png)

Thus,  I think the extent transform should not make null/NaN value the min value.
(This is probably more important than supporting "null/NaN" in the legend.)

**Comment by jheer** [comment]
Thanks! I fixed the `extent` transform to ignore nulls (now published in vega-transforms 1.0.2). That will be included in the next Vega release.

Regarding the other issues, the `bin-ordinal` scale is designed to work with a numerical domain only. I don't expect this to change. For now, I think it is reasonable to map invalid values to gray and omitting a corresponding legend entry.
